json/src/spec/state.rs

```diff
@@ -19,38 +19,74 @@
 use std::collections::BTreeMap;
 use crate::{
 	bytes::Bytes,
-	hash::Address,
+	hash::{Address, H256},
 	spec::{Account, Builtin}
 };
 use serde::Deserialize;
 
+/// Recent JSON tests can be either a map or a hash (represented by a string).
+/// See https://github.com/ethereum/tests/issues/637
+#[cfg(any(test, feature = "test-helpers"))]
+#[derive(Clone, Debug, PartialEq, Deserialize)]
+#[serde(untagged)]
+pub enum HashOrMap {
+	/// When the `postState` is large, tests sometimes just include the state root of the last
+	/// successful block here.
+	Hash(H256),
+	/// The expected `postState` of a test
+	Map(BTreeMap<Address, Account>),
+}
+
+/// Recent JSON tests can be either a map or a string. See https://github.com/ethereum/tests/issues/637
+// todo[dvdplm]: what is the point of doing this for non-test code? Needed?
+#[cfg(not(any(test, feature = "test-helpers")))]
+#[derive(Debug, PartialEq, Deserialize)]
+#[serde(untagged)]
+pub enum HashOrMap {
+	/// When the `postState` is large, tests sometimes just include the state root of the last
+	/// successful block here.
+	Hash(H256),
+	/// The expected `postState` of a test
+	Map(BTreeMap<Address, Account>),
+}
+
 /// Blockchain state deserializer for tests
 #[cfg(any(test, feature = "test-helpers"))]
 #[derive(Clone, Debug, PartialEq, Deserialize)]
 #[serde(deny_unknown_fields)]
-pub struct State(pub BTreeMap<Address, Account>);
+pub struct State(pub HashOrMap);
 
 /// Blockchain state deserializer.
 #[cfg(not(any(test, feature = "test-helpers")))]
 #[derive(Debug, PartialEq, Deserialize)]
 #[serde(deny_unknown_fields)]
-pub struct State(BTreeMap<Address, Account>);
+pub struct State(pub HashOrMap);
 
 impl State {
 	/// Returns all builtins.
 	pub fn builtins(&self) -> BTreeMap<Address, Builtin> {
-		self.0
-			.iter()
-			.filter_map(|(add, ref acc)| acc.builtin.clone().map(|b| (add.clone(), b)))
-			.collect()
+		match &self.0 {
+			HashOrMap::Hash(_) => BTreeMap::default(),
+			HashOrMap::Map(map) => {
+				map.iter().filter_map(|(add, ref acc)| {
```

Comment:
impl `Iterator` instead of having to do this matching every time?

Reply:
So I tried this; compiler yells at me. My attempt:

```
impl Iterator for State {
	type Item = <BTreeMap<Address, Account> as Iterator>::Item;
	fn next(&mut self) -> Option<Self::Item> {
		match self.0 {
			HashOrMap::Map(map) => map.iter().next(),
			HashOrMap::Hash(_) => None
		}
	}
}
```

The error:

```
   --> json/src/spec/state.rs:103:1
    |
103 | impl IntoIterator for State {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<I> std::iter::IntoIterator for I
              where I: std::iter::Iterator;
```

What am I doing wrong?

Reply:
Bad suggestion sorry, don't bother with this I think you need another type to workaround this such as `StateIter` or alternatively `impl IntoIter for &State`